topiary-core/src/atom_collection.rs

```diff
@@ -883,14 +931,138 @@ impl AtomCollection {
+                } else {
+                    atoms_with_comments_before.push_front(atom);
+                }
+            } else {
+                atoms_with_comments_before.push_front(atom);
+            }
+        }
+        let mut blank_line_before_first_comment = false;
+        // If we still have comments left, add them at the beginning of the file
+        while let Some(CommentWithContext {
+            comment,
+            blank_line_after,
+            blank_line_before,
+        }) = comments_queue_with_context.pop_back()
+        {
+            if blank_line_after {
+                atoms_with_comments_before.push_front(Atom::Blankline)
+            } else {
+                atoms_with_comments_before.push_front(Atom::Hardline);
+            }
+            atoms_with_comments_before.push_front(comment);
+            blank_line_before_first_comment = blank_line_before;
+        }
+        if blank_line_before_first_comment {
+            atoms_with_comments_before.push_front(Atom::Blankline);
+        }
+
+        // Second pass: get atoms in reverse order, put comments that come after atoms at the correct position
+        let mut atoms_with_all_comments: VecDeque<Atom> = VecDeque::new();
+        let mut comments_queue: VecDeque<Atom> = VecDeque::new();
+        while let Some(atom) = atoms_with_comments_before.pop_front() {
+            if let Atom::Leaf { id, .. } = atom {
+                while let Some(Comment {
+                    content,
+                    original_column,
+                    // We don't care about blank lines here: they only matter for comments that
+                    // come before atoms.
+                    ..
+                }) = self.comments_after.entry(id).or_default().pop_front()
+                {
+                    let comment_atom = Atom::Leaf {
+                        content,
+                        id: self.next_id(),
+                        original_column,
+                        single_line_no_indent: false,
+                        multi_line_indent_all: true,
+                    };
+                    comments_queue.push_front(comment_atom);
+                }
+            }
+            if Atom::Hardline == atom || Atom::Blankline == atom {
+                // Append the comments, each one preceded by a space
+                while let Some(comment) = comments_queue.pop_back() {
+                    atoms_with_all_comments.push_back(Atom::Space);
+                    atoms_with_all_comments.push_back(comment);
+                }
+            }
+            atoms_with_all_comments.push_back(atom);
+        }
+        // If we still have comments left, add them at the end of the file
+        while let Some(comment) = comments_queue.pop_back() {
+            atoms_with_all_comments.push_back(Atom::Space);
+            atoms_with_all_comments.push_back(comment);
+        }
+
+        self.atoms = atoms_with_all_comments.into()
+    }
```

Comment:
I kind of see what's going on here, but I find it very hard to follow. Your comments help, but I think this big function would benefit from more documentation on the algorithm and the data structures that are being used.